### PR TITLE
Add jq language

### PR DESCRIPTION
**Pull request by wader**
https://stedolan.github.io/jq/

## Description

Language color is based on code example color from jq documentation.

## Checklist:

- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      
      -  https://github.com/search?q=extension%3Ajq&type=Code most of the files found seem to be jq and not jsoniq.
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - https://github.com/stedolan/jq/blob/master/src/builtin.jq
    - Sample license(s):
      - https://raw.githubusercontent.com/stedolan/jq/master/COPYING
      - language-jq submodule is fork of https://github.com/bencampion/language-jq that is under MIT https://github.com/bencampion/language-jq/blob/master/package.json
  
  - [x] I have included a syntax highlighting grammar: https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_format=auto&grammar_url=https%3A%2F%2Fraw.githubusercontent.com%2Fwader%2Flinguist-language-jq%2Fmaster%2Fgrammars%2Fjq.tmLanguage.json&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fraw.githubusercontent.com%2Fstedolan%2Fjq%2Fmaster%2Fsrc%2Fbuiltin.jq&code=
  - [ ] I have updated the heuristics to distinguish my language from others using the same extension.

### CODE REVIEW

**Comment by wader** [comment]
Hi, I will probably need some help with heuristics to distinguish between jsoniq and jq, they both use `.jq`. Also i'm not sure I did the correct thing and added jq as `source.jq.2` as `source.jq` is already used by jsoniq. I also did a fork of https://github.com/bencampion/language-jq to https://github.com/wader/language-jq to be able to get scope name `source.jq.2`.

File: lib/linguist/languages.yml
Comment (wader) [inline lib/linguist/languages.yml:68]: Not sure why this got changed. Think the update-ids script reindented?

**Comment by wader** [comment]
Nice work on the CI system for this repo 👍  catches a lot of things!

File: lib/linguist/languages.yml
Comment (Alhadis) [reply lib/linguist/languages.yml:68]: > Think the update-ids script reindented?

Yes. When it updates `languages.yml`, the entire file is replaced _in situ_ with the YAML-encoded representation of the in-memory database.

Here's what `update-ids` would look like if it were Node.js/JSON instead of Ruby/YAML:

~~~js
#!/usr/bin/env node
import {readFileSync, writeFileSync} from "fs";

const jsonFile = "./lib/linguist/languages.json";
let languages;

// Load database
languages = readFileSync(jsonFile, "utf8");
languages = JSON.parse(languages);

// Generate IDs for any entries that lack one
for(const name in languages){
	if(!("language_id" in languages[name]))
		languages[name].language_id = generateRandomID();
}

// Write updated database back to file
languages = JSON.stringify(languages, null, "\t");
writeFileSync(jsonFile, languages);
~~~

Comment (Alhadis) [reply lib/linguist/languages.yml:68]: So when you updated the IDs, the script naturally reformatted anything that was added manually in the time since `update-ids` was last run.

File: lib/linguist/languages.yml
Comment (Alhadis) [inline lib/linguist/languages.yml:2704]: Shouldn't this be

```suggestion
JQ:
```

instead?

File: lib/linguist/languages.yml
Comment (wader) [reply lib/linguist/languages.yml:68]: Ok! should i revert or keep it in the PR?

Comment (Alhadis) [reply lib/linguist/languages.yml:68]: Keep it in the PR. This happens from time-to-time.

File: lib/linguist/languages.yml
Comment (wader) [reply lib/linguist/languages.yml:2704]: Can do, what is the convention? i think jq commonly is all lowercase but it seems the yml file uses captial keys?

**Comment by wader** [comment]
Just curious, i haven't specified a path to the syntax file in the submodule, will be derived automatically somehow?

File: lib/linguist/languages.yml
Comment (Alhadis) [reply lib/linguist/languages.yml:2704]: If a language's name uses non-standard or unusual casing (as [jq](https://stedolan.github.io/jq/) does), then we use that; otherwise, **<ins>L</ins>abel <ins>C</ins>ase** is used (unless it's an obvious acronym, like HTML).

Aside from renaming the YAML entry, the `samples/Jq` directory will need to be renamed to use the updated capitalisation. Now, committing this might be tricky if you're using an OS with a case-insensitive filesystem (e.g., macOS). Running the following commands from the root of your fork should do the trick:

~~~sh
git config --local --add core.ignorecase false
git add --all samples/Jq samples/JQ
~~~

**Comment by Alhadis** [comment]
> Just curious, i haven't specified a path to the syntax file in the submodule, will be derived automatically somehow?

Aye. [This](https://github.com/github/linguist/blob/b2834449f1dd9dd9b76652092bad9fc593f687d3/tools/grammars/compiler/loader.go#L106-L129) should give you the gist of where Linguist looks for grammar files.

> […] as `source.jq` is already used by `jsoniq`

Are you sure that's for a different format also named JQ? If not, we can use the JSONIQ repository's existing `source.jq` grammar.

File: lib/linguist/languages.yml
Comment (Alhadis) [reply lib/linguist/languages.yml]: Bloody hell, I gave you the wrong capitalisation. 😰 That last line should have been

~~~sh
git mv samples/Jq samples/jq
~~~

So you'll now have to run

~~~sh
git mv samples/JQ samples/jq
~~~

Apologies for the screw-up. 😞

**Comment by wader** [comment]
> > […] as `source.jq` is already used by `jsoniq`
> 
> Are you sure that's for a different format also named JQ? If not, we can use the JSONIQ repository's existing `source.jq` grammar.

They have similar syntax and similar uses so not surprised it's confusing :) It's probably the reason linguist current is mis-detect some jq syntax as JSONIQ.

Compare:
https://github.com/stedolan/jq/blob/master/src/builtin.jq
vs
https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_format=auto&grammar_url=https%3A%2F%2Fraw.githubusercontent.com%2Fwader%2Flinguist-language-jq%2Fmaster%2Fgrammars%2Fjq.tmLanguage.json&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fraw.githubusercontent.com%2Fstedolan%2Fjq%2Fmaster%2Fsrc%2Fbuiltin.jq&code=

So not sure what to do about the scope name, not great. Would be good in the future to be able to point the vendor submodule to a syntax file used for vscode/atom etc but extensions for them seem to use `source.jq` for both jq and JSONIQ.

**Comment by wader** [comment]
Do we have to add some heuristics because they are so similar?

File: lib/linguist/languages.yml
Comment (wader) [reply lib/linguist/languages.yml]: No worries, should be "jq" as yml key also?

**Comment by Alhadis** [comment]
> So not sure what to do about the scope name, not great

I'd submit a PR to the JSONiq repository to change the scope-name to `source.jsoniq`. I don't understand why the author picked `source.jq` as a unique identifier for their JSONiq grammar…

**Comment by Alhadis** [comment]
> Do we have to add some heuristics because they are so similar?

I think so. Try `/^def[ \t]+\w+(?:\(.*?\))?:/` for jq.

**Comment by wader** [comment]
> > So not sure what to do about the scope name, not great
> 
> I'd submit a PR to the JSONiq repository to change the scope-name to `source.jsoniq`. I don't understand why the author picked `source.jq` as a unique identifier for their JSONiq grammar…

Mm 😞  i guess they picked it based on extension name, both use .jq 😄

**Comment by wader** [comment]
Some unique things for JSONIQ, uses `(: ... :)` for comments, `:=` assignment and `{| ... |}` for scopes

**Comment by Alhadis** [comment]
Just make sure the heuristics for each language are matching syntax that's invalid in the other.

Remember, it doesn't have to be foolproof: our new classifier is doing a pretty good job of identifying languages from the `samples/*` it's trained by. :wink:

**Comment by pjeby** [comment]
Yeah, shouldn't the token frequencies be pretty indicative without heuristics?  I mean, you can't write much jq without `def`, and JSONiq comment syntax is invalid syntax in jq.  I think the commonly-used keywords/functions are different in each as well.

**Comment by wader** [comment]
@pjeby @AlaineN ok how to find out if classifier is enough?

Suggest heuristic if needed:
```yml
- extensions: ['.jq']
  rules:
  - language: jq
    pattern: '^def[ \t]+\w+(?:\(.*?\))?:'
  - language: JSONiq
    pattern: '\(:|:\)|:=|{\||\|}'
```

In the patterns does `^` anchor to start of whole source text or line starts?

**Comment by wader** [comment]
@Alhadis Great work, also think it make most sense to rename scope for JSONiq. Should i rename the scope to `source.jq` in https://github.com/wader/linguist-language-jq and update the submodule commit? note that my repo is a fork of https://github.com/bencampion/language-jq but it also has some additional fixes for arrays and keywords, see https://github.com/bencampion/language-jq/issues

**Comment by Alhadis** [comment]
@wader I've submitted a PR to fix the JSONiq grammar's scope. See wcandillon/language-jsoniq#23.

> In the patterns does `^` anchor to start of whole source text or line starts?

In *most* RegExp engines, `^` matches the start of the source-text **unless** multiline mode is active (activated by `/m`). When in multiline mode, `^` will match both the start of a string *and* the source-text (and the converse is also true of `$`). This is complemented by a separate mode called `dotall` (`/s`), which causes `.` to match newlines in addition to non-newlines (by default, `.` won't match a newline character).

Ruby, however, does something of questionable sanity: not only does it enable multiline mode by default, its multiline mode *also* enables `dotall` semantics.

So, to give you the TL;DR answer: yes, `^` matches both newlines and start-of-input. It's worth pointing out that the latter can be explicitly requested by way of the `\A` sequence (which may not behave the way you'd expect if regexes are being tested line-by-line).

**Comment by Alhadis** [comment]
> Should i rename the scope to `source.jq` in https://github.com/wader/linguist-language-jq and update the submodule commit?

Yes, but hold off until @wcandillon approves/merges wcandillon/language-jsoniq#23. In addition to bumping the JSONiq submodule, you'll need to update...

1.	[`grammars.yml`](https://github.com/github/linguist/blob/b2834449f1dd9dd9b76652092bad9fc593f687d3/grammars.yml#L535):
	```diff
	 vendor/grammars/language-jsoniq:
	-- source.jq
	+- source.jsoniq
	```
2.	[`languages.yml`](https://github.com/github/linguist/blob/b2834449f1dd9dd9b76652092bad9fc593f687d3/lib/linguist/languages.yml#L2580)
	```diff
	 JSONiq:
	   color: "#40d47e"
	   type: programming
	   ace_mode: jsoniq
	   codemirror_mode: javascript
	   codemirror_mime_type: application/json
	   extensions:
	   - ".jq"
	-  tm_scope: source.jq
	+  tm_scope: source.jsoniq
	```

**Comment by wader** [comment]
Ok let's wait a bit, thanks for all the help

**Comment by Alhadis** [comment]
> note that my repo is a fork of https://github.com/bencampion/language-jq but it also has some additional fixes for arrays and keywords

If @bencampion isn't maintaining that package/repository anymore, perhaps ask if he'd be willing to transfer ownership of it to you. That way, you can publish improvements that benefit both Linguist *and* Atom users. 👍 Failing that, try submitting a PR. @bencampion appears reasonably active on GitHub, so I doubt contributions will fall on deaf ears.

IMHO, publishing a fork should really only be done when you *know* a grammar repository is dead and unmaintained. Or if Linguist's requirements clash with those of upstream (generally in placement and naming of files, switching to unsupported grammar formats, etc).

**Comment by wader** [comment]
True, maybe i should turn my issues into a PR instead. Sadly it seems like vscode does not have CSON support from what i can tell, but atom and Linguist as JSON support so maybe it should be maintained in that format.

The `source.jq` collision made me do a fork but now i think your approach is better and hopefully we can make the grammar situation a be mess messy.

**Comment by Alhadis** [comment]
<pre language="yaml"><code>- language: JSONiq
  pattern: '<mark><ins>\(:|:\)</ins></mark>|:=|{\||\|}'
</code></pre>

You might want to match `(:…:)` in pairs:

```regexp
(?-m:\(:.*?:\))
```

Otherwise, the heuristic risks matching something like `:)`, which could easily appear alongside a snarky comment:

~~~jsoniq
let $rant := "Back in my day, :) was how you wrote 😄."
    $rant += "Damn kids these days with their newfangled emojis…"
~~~

For performance reasons, I've restricted the regex to single-line searches using `(?-m)`, which disables Ruby's multiline mode (going back to my [earlier comment](https://github.com/github/linguist/pull/5233#issuecomment-789288012) about Ruby's whacky defaults…)

**Comment by pjeby** [comment]
> ```yaml
> pattern: '^def[ \t]+\w+(?:\(.*?\))?:'
> ```

Note that jq is a stream-formatted language (i.e. line feeds and other whitespace are interchangeable), and that `def`  can occur anywhere in a jq pipeline.  e.g. this is valid jq that would not trigger the above heuristic:

```
{x:42} | def x: .x; x
```

**Comment by Alhadis** [comment]
> `def` can occur anywhere in a jq pipeline

That's fine. We don't need heuristics to be exhaustive and airtight, unless we're distinguishing between *very* similar languages (e.g., C / C++ / Objective-C / Objective C++). Otherwise, a heuristic need only match commonly-encountered constructs that cover the lion's share of real-world uses, and do so accurately.

Don't forget, there's still the classifier to fall back on when heuristics (and every other strategy) fails to match a language.

**Comment by wader** [comment]
@Alhadis let me know when i should update things. I guess we should merge a rename and update of JSONiq submodule first?

**Comment by Alhadis** [comment]
@wader See wcandillon/language-jsoniq#24. I couldn't resist fixing some glaring mistakes in the JSONiq grammar (though I left the XQuery grammar alone). We'll wait for it to get merged, then we can bump the grammar. :+1:

**Comment by wader** [comment]
> @wader See [wcandillon/language-jsoniq#24](https://github.com/wcandillon/language-jsoniq/pull/24). I couldn't resist fixing some glaring mistakes in the JSONiq grammar (though I left the XQuery grammar alone). We'll wait for it to get merged, then we can bump the grammar. 👍

Ok 👍 does the jq grammar look ok? i've noticed that it do allow quite a lot of things that are not valid jq, eg `1 || 1` but maybe it's hard to describe syntax exactly.

I noticed a merge commit got pushed, should i stay away from rebasing and just add more commits now?

**Comment by Alhadis** [comment]
> I noticed a merge commit got pushed

Aye, you'll want to update your branch locally by pulling from https://github.com/wader/linguist/tree/jq. 😉

> i've noticed that it do allow quite a lot of things that are not valid jq, eg 1 || 1 but maybe it's hard to describe syntax exactly.

That's fine. 👍 Syntax highlighting isn't the place for validating semantics; only stuff that's clearly illegal syntax really deserves to be marked as an error.

**Comment by wader** [comment]
Created a PR with the fixes https://github.com/bencampion/language-jq/pull/4

**Comment by wader** [comment]
@AlaineN Is linguist used in highlight code in github flavoured markdown?

**Comment by AlaineN** [comment]
> @AlaineN Is linguist used in highlight code in github flavoured markdown?

I think you may be intending to tag someone else.

**Comment by Alhadis** [comment]
> I think you may be intending to tag someone else.

I think I might be that "someone else". 😁 @Alhadis, @AlaineN… close enough for autocomplete to get it wrong. Right, @lildud?

**Comment by Alhadis** [comment]
> ~~\@AlaineN~~ @Alhadis Is linguist used in highlight code in github flavoured markdown?

Alright, two things:

1. Linguist actually doesn't do the highlighting; it simply selects the grammar which GitHub's highlighting engine (something called "PrettyLights") will use to make code look spiffy on GitHub.com
2. I assume that by "GitHub-flavoured markdown", you're referring to fenced code-blocks. In which case, yes, the same grammar is used to highlight embedded code samples. However, this isn't limited to Markdown: other formats supported on GitHub (like [AsciiDoc](https://github.com/Alhadis/Accordion/blob/master/docs/options.adoc)) utilise a similar feature.

File: lib/linguist/languages.yml
Comment (Alhadis) [reply lib/linguist/languages.yml]: Sorry, I missed that last question.

Yes. Everywhere the language's name appears should match the exact characters used in `languages.yml`: Heuristics, tests, the name of the `samples/` subdirectory, the grammar-index, etc.

The only exception to this is the `F*` language, which necessitated a [new field](https://github.com/github/linguist/blob/76c397c7ad07319e766fb1445b8a2307af2f586b/lib/linguist/languages.yml#L3-L5) that allowed us to use filename-safe characters for [`./samples/Fstar/](https://github.com/github/linguist/blob/76c397c7ad07319e766fb1445b8a2307af2f586b/lib/linguist/languages.yml#L1502). However, this is an edge-case, and one you'll almost certainly never have to worry or care about.

Comment (wader) [reply lib/linguist/languages.yml]: Ok, i assumed so, so already changed it

**Comment by wader** [comment]
> > @AlaineN Is linguist used in highlight code in github flavoured markdown?
> 
> I think you may be intending to tag someone else.

Oj sorry for that, not sure how your name got into the auto completer

**Comment by Alhadis** [comment]
> not sure how your name got into the auto completer

She's a GitHub staff member, and Linguist is hosted on the @github org, so... I imagine that's all it takes to be listed as an autocomplete suggestion.

**Comment by wader** [comment]
Getting close to merge? what happens after merge, will take some time until used by official github and jq-files being detected as jq?

**Comment by Alhadis** [comment]
> Getting close to merge? what happens after merge, will take some time until used by official github and jq-files being detected as jq?

Hey @wader, sorry for the slow response.

Yes, this LGTM. 👍 As for merging, changes made to Linguist only take effect on GitHub once there's been a new release of Linguist. Hence you won't see the improved highlighting on the site straight away.

**Review by Alhadis — APPROVED** [review]

**Comment by wader** [comment]
Ok! and as i understand submodules are updated regularly in linguist so no need to send PR if the syntax is updated?

And thanks again for all the help!

**Comment by Alhadis** [comment]
> Ok! and as i understand submodules are updated regularly in linguist so no need to send PR if the syntax is updated?

Grammars are all bumped to the latest versions right before a release is cut. Only in peculiar scenarios (such as this one) is it necessary to manually bump a grammar ourselves.

**Review by lildude — APPROVED** [review]
LGTM. Thanks 🙇

**Pull request merged**
